### DISCUSSION TITLE
Complex quries examples

### DISCUSSION THREAD

**Discussion by subhendukundu** [supabase/supabase#2755] [Questions] [answered]
Hey team! Trying to migrate from Hasura to Supabase. So I need a bit of help in terms of best practices. 

For example, if I had to get a post's, like count with first 10 liked user including if the current loggedin user liked it.

I have done it hasura like so
```
query {
  post_by_pk(project_id: "some_uuid_for_post"){
    ifLiked: likes(where: {
      user_id: {
        _eq: "i_logged_in_with_this_uuid"
      }
    }) {
      id
    }
    likes(limit: 10) {
      user {
        name
        id
      }
    }
    total_like: likes_aggregate {
      aggregate {
        count
      }
    }
  }
}
```

Also would love to see some of the usecase examples in docs or examples for quick understandings.

**Comment by dshukertjr** [comment]
Hi @subhendukundu!

For complex queries like this, using stored procedures would be the route to take. 

It's not the same exact scenario as your case, but I have created something similar to your situation where I wanted to get the like count, comment count, whether the user have liked it or not and stuff.

```sql
create or replace function get_video_detail(video_id uuid, user_id uuid)
returns table(id uuid, url text, image_url text, thumbnail_url text, gif_url text, created_at timestamptz, description text, user_id uuid, user_name text, user_description text, user_image_url text, location text, like_count int, comment_count int, have_liked int)
as
$func$
    select
        videos.id,
        videos.url,
        videos.image_url,
        videos.thumbnail_url,
        videos.gif_url,
        videos.created_at,
        videos.description,
        users.id as user_id,
        users.name as user_name,
        users.description as user_description,
        users.image_url as user_image_url,
        st_astext(videos.location) as location,
        (select count(*) from likes where video_id = videos.id)::int as like_count,
        (select count(*) from comments where video_id = videos.id)::int as comment_count,
        (select count(*) from likes where video_id = videos.id and user_id = $2)::int as have_liked
    from videos
    join users on videos.user_id = users.id
    where videos.id = $1;
$func$
language sql;
```

You can call this like this:
```typescript
const {data, error} = await supabase.rpc('get_video_detail', {video_id, user_id});
```

>Also would love to see some of the usecase examples in docs or examples for quick understandings.

100% agree. It would be awesome if we had a section within the doc with bunch of complex examples of views and stored procedures. 

## Edited
@steve-chavez's answer is amazing. Did not know that this is possible with only JS client. Learning new things about Supabase every single day!

**Reply by donPuerto** [reply]
@dshukertjr Where we can see this after we run?

**Reply by dshukertjr** [reply]
@donPuerto 

You can see it at `database > Functions`
<img width="1043" alt="Screen Shot 2021-08-12 at 8 15 24" src="https://user-images.githubusercontent.com/18113850/129115687-c9d4f500-7251-4a6b-806e-626d4c481437.png">

**Reply by donPuerto** [reply]
Thanks

**Comment by steve-chavez** [comment, marked as answer]
Hey @subhendukundu,

This seems possible with just the JS client. Not sure if I'm getting your tables right(sample DDL would help) but I came up with this:

```js
const { data, error } = await supabase
  .from('posts')
  .select(`
    *,
    ifliked:likes(id),
    likes(user(name,id)),
    total_like:likes(count)
  `)
  .eq('user_id', "some_uuid_for_post")
  .eq('ifliked.user_id', "i_logged_in_with_this_uuid")
  .limit(10, { foreignTable: 'likes' });
```

For more complex queries, do follow @dshukertjr approach in the other answer.
